h/traversal/roots.py

```diff
@@ -0,0 +1,221 @@
+https://docs.pylonsproject.org/projects/pyramid/en/latest/api/config.html#pyramid.config.Configurator.add_route
+
+The intended pattern in this app is that all root resources **should return
+context objects** from :py:mod:`h.traversal.contexts` (or raise
+:py:exc:`KeyError`), they shouldn't return other types of object (e.g. they
+shouldn't return model objects directly).
+
+.. note::
+
+   Technically the *classes* in this module are Pyramid "root factories"
+   (hence the ``factory`` argument to :py:func:`pyramid.config.Configurator.add_route`)
+   and the *object instances* of these classes are the Pyramid "root resources"
+   that the factories return when called (instantiated).
+
+.. note::
+
+   In order to encapsulate SQLAlchemy in the models and services, root
+   resources should look up objects in the DB by calling a ``@classmethod`` of
+   a :py:mod:`h.models` class or a method of a service from
+   :py:mod:`h.services`, rather than by doing DB queries directly.
+
+.. seealso::
+
+   The Pyramid documentation on traversal:
+
+   * https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/hellotraversal.html
+   * https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/muchadoabouttraversal.html
+   * https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/traversal.html
+
+"""
+from __future__ import unicode_literals
+
+
+from pyramid.security import (
+    ALL_PERMISSIONS,
+    DENY_ALL,
+    Allow,
+)
+from sqlalchemy.orm import exc
+
+from h import storage
+from h.models import AuthClient
+from h.models import Group
+from h.models import Organization
+from h.auth import role
+from h.interfaces import IGroupService
+from h.traversal.contexts import AnnotationContext
+
+
+class Root(object):
+    """This app's default root factory."""
+    __acl__ = [
+        (Allow, role.Staff, 'admin_index'),
+        (Allow, role.Staff, 'admin_groups'),
+        (Allow, role.Staff, 'admin_mailer'),
+        (Allow, role.Staff, 'admin_organizations'),
+        (Allow, role.Staff, 'admin_users'),
+        (Allow, role.Admin, ALL_PERMISSIONS),
+        DENY_ALL
+    ]
+
+    def __init__(self, request):
+        self.request = request
+
+
+class AnnotationRoot(object):
+    """Root factory for routes whose context is an :py:class:`h.traversal.AnnotationContext`."""
+    def __init__(self, request):
+        self.request = request
+
+    def __getitem__(self, id):
+        annotation = storage.fetch_annotation(self.request.db, id)
+        if annotation is None:
+            raise KeyError()
+
+        group_service = self.request.find_service(IGroupService)
+        links_service = self.request.find_service(name='links')
+        return AnnotationContext(annotation, group_service, links_service)
+
+
+class AuthClientRoot(object):
+    """
+    Root factory for routes whose context is an :py:class:`h.traversal.AuthClientContext`.
+
+    FIXME: This class should return AuthClientContext objects, not AuthClient
+    objects.
+
```

Comment:
@lyzadanger ^ to your previous comment. This is a useful FIXME for example pointing out a non-ideal aspect of the current code.